Mathlib/Data/Nat/ModEq.lean

```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
     fun h => Or.elim h odd_of_mod_four_eq_one odd_of_mod_four_eq_three⟩
 #align nat.odd_mod_four_iff Nat.odd_mod_four_iff
 
+lemma mod_eq_of_modEq {a b n} (h : a ≡ b [MOD n]) (hb : b < n) : a % n = b := by
+  simpa [show a % n = b % n from h] using mod_eq_of_lt hb
+
+variable {ι : Type*}
+open BigOperators
+
+lemma modEq_list_prod_iff {a b} {l : List ℕ} (co : l.Pairwise Coprime) :
+    a ≡ b [MOD l.prod] ↔ ∀ i, a ≡ b [MOD l.get i] := by
+  induction' l with m l ih
+  · simp [modEq_one]
+  · have : Coprime m l.prod := coprime_list_prod_right_iff.mpr (List.pairwise_cons.mp co).1
+    simp [← modEq_and_modEq_iff_modEq_mul this, ih (List.Pairwise.of_cons co)]
+    constructor
+    · rintro ⟨h0, hs⟩ i
+      cases i using Fin.cases <;> simp [h0, hs]
+    · intro h; exact ⟨h 0, fun i => h i.succ⟩
+
+lemma modEq_list_prod_iff' {a} {s : ι → ℕ} {l : List ι} (co : l.Pairwise (Coprime on s)) :
+    a ≡ b [MOD (l.map s).prod] ↔ ∀ i ∈ l, a ≡ b [MOD s i] := by
+  induction' l with i l ih
+  · simp [modEq_one]
+  · have : Coprime (s i) (l.map s).prod := coprime_list_prod_right_iff.mpr
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    simp [← modEq_and_modEq_iff_modEq_mul this, ih (List.Pairwise.of_cons co)]
+
+variable (a s : ι → ℕ)
+
+/-- The natural number less than `(l.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ l`. -/
+def chineseRemainderOfList : (l : List ι) → l.Pairwise (Coprime on s) →
```

Comment:
Why is the return type `{ k // ∀ i ∈ l, k ≡ a i [MOD s i] }`?
I would guess that this is easier to work with as just returning `k : ℕ`, and then having a separate proof, called `chineseRemainderofList_spec` or something, that when `k` is the output, `∀ i ∈ l, k ≡ a i [MOD s i]`.

Reply:
I defined like this because I followed the definition of [Nat.chineseRemainder](https://leanprover-community.github.io/mathlib4_docs//Mathlib/Data/Nat/ModEq.html#Nat.chineseRemainder). It may be a little unnatural, but I preferred the consistency of the definition.

Reply:
If that's how the library already is, it seems fine to me.